apps/editing-toolkit/editing-toolkit-plugin/block-patterns/class-block-patterns.php

```diff
@@ -82,75 +97,50 @@ public function register_patterns() {
 	 *
 	 * @return array
 	 */
-	public function get_patterns() {
-		$patterns_dir = __DIR__ . '/patterns/';
-		$patterns     = array();
-
-		if ( ! is_dir( $patterns_dir ) ) {
-			return $patterns;
-		}
-
-		$directory_handle = opendir( $patterns_dir );
-		if ( ! $directory_handle ) {
-			return $patterns;
-		}
-
-		// Get all the pattern files.
-		$pattern       = readdir( $directory_handle );
-		$pattern_files = array();
-		while ( false !== $pattern ) {
-			// Only allow files ending in .json or .php.
-			if ( substr( $pattern, -5 ) === '.json' || substr( $pattern, -4 ) === '.php' ) {
-				$pattern_files[] = $pattern;
+	private function get_patterns() {
+		$block_patterns = get_transient( $this->patterns_cache_key );
+
+		// Load fresh data if we don't have any patterns.
+		if ( false === $block_patterns || ( defined( 'WP_DISABLE_PATTERN_CACHE' ) && WP_DISABLE_PATTERN_CACHE ) ) {
+			$request_url = add_query_arg(
+				array(
+					'site'     => $this->source_site_id,
+					'language' => $this->get_iso_639_locale(),
+				),
+				'https://public-api.wordpress.com/rest/v1/ptk/patterns'
+			);
+
+			$response = wp_remote_get(
+				esc_url_raw( $request_url ),
+				array( 'timeout' => 20 )
+			);
+
+			if ( 200 !== wp_remote_retrieve_response_code( $response ) ) {
+				return array();
 			}
-
-			$pattern = readdir( $directory_handle );
+			$block_patterns = json_decode( wp_remote_retrieve_body( $response ), true );
+			set_transient( $this->patterns_cache_key, $block_patterns, DAY_IN_SECONDS );
```

Comment:
@glendaviesnz I think this means that the cache will be busted if either the plugin is updated _or_ a day has elapsed since the transient was created. I guess the only risk here is if we have an error in a pattern and we want to easily refresh it across all sites. There's the `WP_DISABLE_PATTERN_CACHE` constant above, but setting that won't just invalidate a single time, but will switch the invalidation off altogether. Do we need a simple way to flush the cache _once_ but not switch off the caching altogether?

Reply:
ah, thanks, overlooked the day flag in the parameter setting 🤦 It would be nice if there was some way we could initiate a cache flush without having to deploy a change. Could the patterns api have an endpoint that returns a segment of the cache key to use in place of `PLUGIN_VERSION`, eg, `PATTERNS_VERSION` which comes from a blog option on the patterns site, then if we increment that the caches get invalidated ?

Reply:
Oh, that's a really interesting idea! Yes, I agree I think it's important that we can come up with a way to initiate cache invalidation from the Patterns Toolkit side of things. I'm currently looking at that for the Patterns endpoint itself, but I like this idea of having an additional flag that we can set _somewhere_ in wpcom, and then the consumer automatically picks it up. It'd be great if we can still work out a way that we're not re-introducing an API call on every server-side request (this time to check if we should make another call), so we might need another simple layer. Perhaps we cache the `PATTERNS_VERSION` key (or whatever we call it) for 5 minutes, so we're calling it fairly frequently, but not on every request?

I'll do some more thinking while I'm experimenting with the API caching. Let me know if you have any more ideas, though, it'd be good to get this right. We can also discuss on pbxlJb-oL-p2#comment-515 if there's stuff that goes deeper into API territory outside of this PR 😀

Reply:
Do you think this is a blocker for this PR? I'd imagine we could easily introduce this in a subsequent update?

Reply:
I would think this could be a follow PR once we have decided how best to handle this.

Reply:
Yep, adding this in a follow-up sounds fine to me, too!